### PR TITLE
Make Listener interface closeable

### PR DESCRIPTION
**Pull request by NatashaL**

### CODE REVIEW

File: flo-runner/src/main/java/com/spotify/flo/context/ChainedListener.java [hunk 1]
```diff
@@ -61,4 +64,29 @@ private void guardedCall(Runnable call) {
       logging.exception(t);
     }
   }
+
+  @Override
+  public void close() throws IOException {
```
Comment (fabriziodemaria) [inline flo-runner/src/main/java/com/spotify/flo/context/ChainedListener.java]: Can we add a test around this?

File: flo-runner/src/main/java/com/spotify/flo/context/ChainedListener.java [hunk 2]
```diff
@@ -61,4 +64,29 @@ private void guardedCall(Runnable call) {
...
+    final Optional<IOException> exception =
+        Stream.of(first, second)
+            .map(this::close)
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .findFirst();
+
+    // TODO: Throw a composition of two exceptions when both first and second throw exceptions
```
Comment (fabriziodemaria) [inline flo-runner/src/main/java/com/spotify/flo/context/ChainedListener.java]: Is it intended to do this in another PR?

File: flo-workflow/src/main/java/com/spotify/flo/context/InstrumentedContext.java [hunk 3]
```diff
@@ -49,7 +50,7 @@
   /**
    * A listener for instrumented evaluation. See {@link InstrumentedContext} for more details.
    */
-  public interface Listener {
+  public interface Listener extends Closeable {
```
Comment (honnix) [inline flo-workflow/src/main/java/com/spotify/flo/context/InstrumentedContext.java]: Having a default implementation in `Listener` which does nothing?

Comment (NatashaL) [reply flo-workflow/src/main/java/com/spotify/flo/context/InstrumentedContext.java]: We can only override the declaration of the `close()` method, not provide an empty implementation, due to Listener being an interface itself. But it seems unnecessary. Whoever implements `Listener` in their implementation class they will be required to implement the `close()` method too.

File: flo-runner/src/main/java/com/spotify/flo/context/ChainedListener.java (hunk 1 above)
Comment (NatashaL) [reply flo-runner/src/main/java/com/spotify/flo/context/ChainedListener.java]: Yes, see 3e7083f

File: flo-workflow/src/main/java/com/spotify/flo/context/InstrumentedContext.java (hunk 3 above)
Comment (honnix) [reply flo-workflow/src/main/java/com/spotify/flo/context/InstrumentedContext.java]: That you can have for an interface.

Comment (NatashaL) [reply flo-workflow/src/main/java/com/spotify/flo/context/InstrumentedContext.java]: Right, yes, we can do that.

**Comment by bergman** [comment]
squashed and merged in 0600752
